---
title: "Gene plot of SCHLAP1 (lncRNA associated with prostate cancer progression) - Figure 3"
output: html_notebook
---

#LOAD LIBRARIES
```{r}
library(Gviz)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(EnsDb.Hsapiens.v86)
library(tidyverse)
```



#READ IN TARGET GENE START AND STOP
```{r}
GOI <- "SCHLAP1"
chr <- "chr2"


target <- EnsDb.Hsapiens.v86 %>% ensembldb::filter(~ symbol == GOI) %>% genes()
target <- target[1]


```


#READ IN DSS METHYLATION REGIONS
```{r}


dss <- read_tsv(file = "/Users/ds/Desktop/projects/mCRPC_lncRNA_analysis//results/DSS_mCPRC_PRAD.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))
dss <- dss %>% dplyr::filter(chrom == chr) %>% dplyr::select(-Stat) %>% mutate(Diff = Diff * 100) %>% GRanges()



#Create DMR tracks
dTrack <- DataTrack(dss, name = "Methylation", background.title = 'purple4', fill = 'purple4')
#plotTracks(list(dTrack), from = start(target), to = end(target), type = c("histogram"), ylim = c(-30,0))

```



#DOWNLOAD TRANSCRIPT TRACKS
```{r}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene %>% exons(columns=c("GENEID", "EXONID", "TXNAME" ))
txdb <- S4Vectors::expand(txdb, c("GENEID", "EXONID", "TXNAME"))
# seqlevelsStyle(target) <- "UCSC"
# gene <- subsetByOverlaps(x = txdb, ranges = target) %>% data.frame()
```


#SUBSET TRACKS FOR SCHLAP1
```{r}
gene <- txdb[txdb$GENEID == 101669767] %>% data.frame()
colnames(gene) <- c("chromosome", "start", "end", "width", "strand", "gene", "exon", "transcript")


#Download HMR/AR chip/ H3k27ac chip bed files
hmr <- read_tsv(file = "/Users/ds/Desktop/projects/mCRPC_lncRNA_analysis/data/hmr.bed", col_names  = c("chrom", "start", "end"))
hmr <- hmr %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()

metarbs <- read_tsv(file = "/Users/ds/Desktop/projects/mCRPC_lncRNA_analysis/data/AR.chip.hg38.bed", col_names = c("chrom", "start", "end"))
metarbs <- metarbs %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()

h3ac <- read_tsv(file = "/Users/ds/Desktop/projects/mCRPC_lncRNA_analysis/data/H3K27ac.chip.hg38.bed", col_names = c("chr", "start", "end"))
h3ac <- h3ac %>% dplyr::select(c(chr, start, end)) %>% GRanges()
h3ac <- subsetByOverlaps(x = h3ac, ranges = GRanges(dplyr::filter(linc, gene.name == "SCHLAP1")))

```


#GENE PLOT
```{r}


#Plot
grtrack <- GeneRegionTrack( genome = "hg38",range = gene, chromosome = chr , name = GOI, transcriptAnnotation = "transcript", background.panel = "lightyellow", background.title = 'navy', fill = 'navy', stacking = "squish", col = 'navy')
atrack <- AnnotationTrack(hmr, name = "HMR", background.title = 'darkred',  background.panel ="lightgrey", fill ='darkred', col = 'darkred')
arbtrack <- AnnotationTrack(metarbs, name = "ARBS", background.title = 'darkgreen' , background.panel = "lightgrey", fill = 'darkgreen', col = 'darkgreen')
atactrack <- AnnotationTrack(h3ac, name = "H3K27ac", background.title = 'black' , background.panel = "lightgrey", fill = 'black', col = 'black')
gtrack <- GenomeAxisTrack()
#itrack <- IdeogramTrack(genome = "hg38", chromosome = chr)



pdf("/Users/ds/Desktop/plot.pdf", width = 10, height = 5)
plotTracks(list(gtrack, dTrack,arbtrack, atactrack, atrack, grtrack), from = start(target), to = end(target), type = "histogram",ylim = c(-30,0), cex.title = 1.5, sizes=c(1,2,1,1,1,1), title.width = 4, rot.title = 0)
dev.off()   
           


```



