---
title: "Differential expression analysis of lncRNAs in primary prostate cancer v.s mCRPC - Figure 3"
output: html_notebook
note: "Refer to figure 1 (preprocess_scrna.Rmd) script for lncRNA sets"
---

#LOAD LIBRARIES
```{r}


library(TCGAWorkflowData)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(edgeR)
library(EnhancedVolcano)
library(Seurat)
library(lumi)
library(tidyverse)
library(ggrepel)
options(ggrepel.max.overlaps = Inf)

```


#DOWNLOAD TCGA PRAD SAMPLE LIST (N=333)
```{r}
#Get prad metadata from TCGA API
mrna_query <- GDCquery(project = "TCGA-PRAD",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
prad_meta<- mrna_query$results[[1]] %>% dplyr::filter(sample_type == "Primary Tumor")
#GDCdownload(mrna_query, method = "api", directory = "/Users/ds/Desktop")
```



#PROCESS PRAD COUNTS
```{r}
#Path for downloaded TCGA PRAD data
full_path <- "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/"
#Merge counts per sample
prad_rna <- do.call(cbind, lapply(list.files(path = full_path), FUN = function(f){
  print(paste0("processing: ",list.files(path = paste0(full_path, f))))
  counts <- read_tsv(file = list.files(path = paste0(full_path, f), full.names = T), skip = 6, col_names = F )
  
  return(data.frame(f = counts[,4]))
  
  
}))
colnames(prad_rna) <- list.files(path = full_path)
#Append gene symbols
symbols <- read_tsv(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/0007888f-8d96-4c01-8251-7fef6cc71596/88215dd0-5841-44f1-9393-eefd8238cbb3.rna_seq.augmented_star_gene_counts.tsv", skip = 6, col_names = F)[,2]
prad_rna$Gene <- symbols$X2

#Write to disk
write.table(x = prad_rna, file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", quote = F, sep = "\t", row.names = F, col.names = T)
```



#FORMAT PRAD COUNT DATA
```{r}

prad_rna <- read.table(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", header = T)
prad_rna <- prad_rna[!duplicated(prad_rna$Gene),]
row.names(prad_rna) <- prad_rna$Gene 
prad_rna <- dplyr::select(prad_rna, -Gene)
colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "^X*", replacement = "")
colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "\\.", replacement = "-")
prad_rna <- prad_rna %>% t() %>% data.frame()#Counts for DEG analysis
prad_rna <- prad_rna %>% mutate(id = row.names(.), batch = "PRAD") %>% merge(dplyr::select(prad_meta , c(id, cases.submitter_id) ))


#Add PRAD purity from GDC website - https://gdc.cancer.gov/about-data/publications/prad_2015
anno_prad <- read.csv(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_supp.csv", header = T)#Supp table 1
anno_prad <- anno_prad %>% dplyr::select(avgRNA_purity, PATIENT_ID, Reviewed_Gleason_sum) #Add tumor purity (RNA) and gleason score
prad_rna <- inner_join(prad_rna, anno_prad, by = c("cases.submitter_id" = "PATIENT_ID")) %>% mutate(Purity = avgRNA_purity * 100, cases = cases.submitter_id, Grade_Mutational_status = Reviewed_Gleason_sum)

```




#FORMAT MCRPC DEEP RNA-SEQ COUNT DATA (N=74)
```{r}
#Download meta data for mCRPC deep RNA-seq
anno_mcrpc <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)
anno_mcrpc <- anno_mcrpc %>% dplyr::select(c(sample_id, wgs_id, biopsy_site,tumor_purity_wgs, tumor_purity_rna , tumor_purity_hist, AR))
anno_mcrpc <- anno_mcrpc %>% mutate(cases =  sample_id, Purity = tumor_purity_rna, Grade_Mutational_status =AR) %>% dplyr::select(cases, Purity, Grade_Mutational_status)#Select tumor purity (RNA) and AR region mutation status


#Read in count data from Featurecounts for mCRPC deep RNA-seq
mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
mcrpc_rna <- mcrpc_rna[!duplicated(mcrpc_rna$Genename),]
row.names(mcrpc_rna) <- mcrpc_rna$Genename
mcrpc_rna <- mcrpc_rna[,-1]
mcrpc_rna <- mcrpc_rna %>% t() %>% data.frame()#Counts for DEG analysis
row.names(mcrpc_rna) <- gsub("\\.", "-",  x = row.names(mcrpc_rna))
mcrpc_rna <- mcrpc_rna %>% mutate(cases = row.names(.), batch = "mCRPC")

#Add mCRPC purity
mcrpc_rna <- mcrpc_rna %>% merge(anno_mcrpc)


```

#MERGE MCRPC AND PRAD COUNT DATASETS USING COMMON GENES
```{r}
#Get genes in common + cases + purity + mutational status
query <- intersect(colnames(prad_rna), colnames(mcrpc_rna))

#Combine PRAD and mCRPC - rows are samples and columns are genes + RNA purity + mutational status/gleason score
rna <- rbind(prad_rna[,query], mcrpc_rna[,query])



```


#EDGER DIFFERENTIAL GENE EXPRESSION ANALYSIS
```{r}
#Extract count data
counts <- dplyr::select(rna, -c(cases, Purity, batch, Grade_Mutational_status)) %>% t() %>% data.frame()
colnames(counts) <- rna$cases

#Create dge object - create counts matrix and sample anno df with library sizes
x <- DGEList(counts = counts)

#Add cluster variable to designate group membership to sample df
x$samples <- cbind(x$samples, dplyr::select(rna, c( Purity, batch )))

#Remove genes with low expression across samples
keep.exprs <- filterByExpr(x, min.count = 5)#Use min count of 5
x <- x[keep.exprs,, keep.lib.size = F]

#Calculate effective library size using EdgeR TMM method
x <- calcNormFactors(x, method = "TMM")

#Create design matrix (w/o intercept) specifying group membership and tumor purity as covariate
design <- model.matrix(~ 0 + batch + Purity, data = rna) 
colnames(design) <- c("mCRPC", "PRAD", "Purity")

# #Create contrasts to calculate
contr.matrix <- makeContrasts(mCRPC - PRAD, levels = colnames(design))


#Normalize counts to log CPM using voom
v <- voom(x, design = design)

#Fit LM according to design
vfit <- lmFit(v, design)

#Calculate contrasts i.e logFC for genes across groups
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

#Calculate t-stats and p vals using Emp bayes method
efit <- eBayes(vfit)


#Get list of all genes, arrange in decreasing order
bulk_deg <- topTable(efit, number = Inf, coef = 1) %>% data.frame(Genes = row.names(.))  %>% arrange(-logFC) 
bulk_deg <- bulk_deg %>% arrange(-logFC)



```

#PLOTTING DEG RESULTS
```{r}
#Merging DEG results with lncRNA markers
bulk_deg <- bulk_deg %>% left_join( y = final_markers, by = c("Genes" = "gene")) %>% mutate(Cluster = ifelse(is.na(cluster), "Unknown", cluster))
bulk_deg <- bulk_deg %>% mutate(Alpha = ifelse(Cluster == "Unknown", 0.1, 1))

                                          
#Volcano plot                                   
plot <- ggplot(bulk_deg, aes(x = logFC, y = -log10(adj.P.Val), color = Cluster, alpha = Alpha, label = Genes) ) + geom_point() + scale_color_manual(values = c("purple", "orange", "red", "black", "darkgreen", "navy", "grey")) + geom_hline(yintercept = -log10(0.1), linetype= 'dashed') + geom_vline(xintercept = c(-1,1), linetype = 'dashed') + geom_text_repel(data = dplyr::filter(bulk_deg, abs(logFC) > 1 & adj.P.Val < 0.1  & Cluster !="Unknown" ), force = 0.8, size = 2) + scale_alpha(range = c(0.8,1)) + theme_classic()
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)


#Write results to disk
#write.table(x = bulk_deg, file = "/Users/ds/Desktop/projects/lncRNA/results/mCRPC_PRAD_deg.txt", quote = F, sep = "\t", row.names = F, col.names = T)
write.csv(x = bulk_deg, file = "/Users/ds/Desktop/Supplementary.Table2.csv",quote = F, row.names = F)




```


#SUBSET UP AND DOWN REGULATED LNCRNAS
```{r}

#Read in bulk DEG results between mCRPC and PRAD
bulk_deg <- read.table(file = "/Users/ds/Desktop/projects/mCRPC_lncRNA_analysis/results/mCRPC_PRAD_deg.txt", header = T, sep = "\t")


#Map up and down regulated genes to clusters
up <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC > 1 & Cluster != "Unknown") # FDR 10% and abs logFC > 1
down <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC < -1 & Cluster != "Unknown")  
bulk_deg <- bulk_deg %>% left_join( y = final_markers, by = c("Genes" = "gene")) %>% mutate(Cluster = ifelse(is.na(cluster), "Unknown", cluster))
bulk_deg <- bulk_deg %>% mutate(Alpha = ifelse(Cluster == "Unknown", 0.1, 1))


```


#METHYLATION ANALYSIS
```{r}


#Read in DMRs from DSS in Zhou et al. 2020 Nature Genetics between mCRPC and PRAD
dss <- read_tsv(file = "/Users/ds/Desktop/projects/mCRPC_lncRNA_analysis//results/DSS_mCPRC_PRAD.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))
dss <- GRanges(dss)
regions <- GRanges(linc) #Convert lncRNAs to GRanges

#Subset DMR (differentially methylated regions) nearby up- and down-regulated genes
dss <- data.frame(linc, dss[nearest(x =  GRanges(regions), subject = dss),] ) %>% inner_join(bulk_deg, by = c("gene.name" = "Genes" )) %>% mutate(Sig = ifelse(gene.name %in% up$Genes & Diff < -0.05, "A", ifelse(gene.name %in% down$Genes & Diff > 0.05, "B", "Other")   ))

#Plotting - DMR cutoff of 5% and logFC cutoff of 1
plot <- dss %>% ggplot(aes(x = Diff * 100, y = logFC, color = Cluster, label= gene.name, alpha = Sig)) + geom_point(size= 1) + geom_vline(xintercept = c(-5,5), linetype = "dashed")  + geom_hline(yintercept = c(-1,1), linetype = "dashed") + geom_text_repel(data = dplyr::filter(dss, Sig %in% c("A", "B") & Cluster != "Unknown"), force =10, size = 5) + scale_alpha_manual(values = c(1,1,0.5))  + ylab("logFC mCRPC v.s. PRAD") + xlab("Methylation difference mCPRC - PRAD") + scale_color_manual(values = c("purple", "orange", "red", "black", "darkgreen", "navy", "grey")) + theme_classic() + theme( text = element_text(size = 25))
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 15, height = 7)
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10, height = 10)

plot
```


